### ISSUE TITLE
[FEATURE REQUEST] Backport observability improvements to 1.11.x

### ISSUE THREAD

**Issue by stefanobaghino-da** [digital-asset/daml#11604]
I confirm that, if this is a feature request that has security implications, I already contacted security@digitalasset.com and followed the [responsible disclosure policy](https://digitalasset.com/security/).

I confirm that this is not a question or a request for technical support by the community, for which the [Daml forum](https://discuss.daml.com/) is available.

**What is the problem you want to solve?**
Identifying performance issues is very difficult with the limited observability provided as part of 1.11.x. In order to improve that, we would benefit from backporting logging and metrics collection improvements that have happened since 1.11.0 was released.

**What is the solution you would propose?**
Backport the logging improvements that have been performed on `main` to `release/1.11.x`.

**Describe alternatives you've considered**
Unfortunately there are no alternatives to have reliable feedback about the status of the application.

**Comment by stefanobaghino-da** [comment]
@realvictorprm

1. please list in a single comment all the improvements that have been made since 1.11.0 was release (this happened before you joined DA and worked on those improvements yourself)
2. please evaluate what performance number are recorded for individual requests
   * we need to be able to identify performance hot spots, so we would need to track for each request:
     * what query has been issued
     * how long it took to process on the JSON API server
     * how long it took to come back from the database

**Comment by realvictorprm** [comment]
OK after succesfully failing in having any bash skill I managed to find some of my bash skills :joy: Regarding metrics & logging I found these most likely relevant changelog entries (with commit hash):
```
814813725940a2646392f39db6ebaaf6549c0327

- [JSON-API] Connection tries from the json api to the ledger now include the logging context, more specifically the instance_uuid is included in each logging statement.

80b65f6def1aa6e65724d8a34cc92b6ef1e21c11

- [JSON-API] You can now find a section `Metrics` in the http-json api documentation explaining how to enable metrics and which are available

e4585295c6714524d61732f1211d20e461be635d

- [JSON-API] Timing metrics are now available for party management, package management, command submission and query endpoints.
- [JSON-API] Also added a timing metric for parsing and decoding of incoming json payloads

13767b5c5d531690a0d4bb7e77cf2a66ab9ac3d3

[HTTP-JSON]
- fixed that json log output could not be enabled via cli options except via usage of env vars

beb2d138f2939f3d9cfbb29cdf6a607f9df6da2b

- [Ledger HTTP Json Service] Logging output can now be in JSON either via providing the cli option `--log-encoder json` or via setting the env var `LOG_FORMAT_JSON=true`

4f0ceff7a2ad2865087250a7f98e17e8fd7e8f75

http-json:
- add contextual id as logging context to distinguish different application runs in logs
- add request id as logging context to distinguish different http requests within an application run
- add for non-static endpoints trace logs which show how long processing it took in ns

```

However I'm currently not completely convinced that these are all relevant commits. Overall it should however at least cover e.g.:
- logging incoming requests if the log level is debug
- overall metrics for the endpoints

But I think we are currently _not_ covering _logging_ the duration of a query. We do collect _metrics_ for that though (as far as I can remember).

**Comment by realvictorprm** [comment]
Noticed that my search contained a small failure which causes a bunch of commits to be not included. Updated list (which I hope is more correct):
```
f4df1cc4d06eeb2c1de2677c35b5aa9bfb84c9fd

- [JSON API] Several kinds of gRPC server errors are now reported with
  associated HTTP statuses; for example, a Daml-LF interpreter error now
  returns a 400 instead of a 500, and an exercise on an archived contract
  returns a 409 Conflict instead of a 500.  Errors internal to JSON API
  (e.g. internal assertion failures) are no longer detailed in the HTTP
  response; their details are only logged.
  See `issue #11184 <https://github.com/digital-asset/daml/pull/11184>`__.

6492ceb93304e23a1f94ae34c3be612bf3b7c98d

[JSON API] Ledger connection errors are now logged at every attempt

814813725940a2646392f39db6ebaaf6549c0327

- [JSON-API] Connection tries from the json api to the ledger now include the logging context, more specifically the instance_uuid is included in each logging statement.

80b65f6def1aa6e65724d8a34cc92b6ef1e21c11

- [JSON-API] You can now find a section `Metrics` in the http-json api documentation explaining how to enable metrics and which are available

6abb903a7f370eee0834a39b62b33028107f9745

- [JSON-API] Logging of the request and response bodies are now available for appropriate requests if the chosen log level is equal or lower than DEBUG. These can then be found in the logging context of the request begin & end log messages (The field names in the ctx are "request_body" and "response_body").

b59f36591d34c701674f0f66b9cdb3998274a0d1

- [JSON-API] If the service is put behind a proxy filling either of these headers X-Forwarded-For & X-Real-Ip then these will now be respected for logging the request source ip/url

7521ec998807f9093c7fe4653d8ee292441a7968

- [JSON-API] The contract id is now also included in the logging ctx when logging exercise commands

1f5aa4491cebda97568f1f1358e81587b4afbf93

- [JSON-API] Timing metrics which measure how long the processing of a command submission request takes on the ledger are now available

0391f3529f2f3f113839427f152220f4ce4b1ca3

- [JSON-API] The database operations (regardless of in-memory or postgres) contain now metrics for fetching contracts by id or key (seperate metrics foreach)
- [JSON-API] The time for a repsonse payload construction of a request is now also tracked in a metric

e4585295c6714524d61732f1211d20e461be635d

- [JSON-API] Timing metrics are now available for party management, package management, command submission and query endpoints.
- [JSON-API] Also added a timing metric for parsing and decoding of incoming json payloads

1852830833e124fbf0e2985e412e50ca8725115e

- [JSON-API] The metrics which describe the amount of these concurrent events is now available:
	- running http request
	- command submissions
	- package allocations
	- party allocations

cfee5a2d8b452954830e0a065763e114e5ec2b78

- [JSON API] For applicable requests actAs, readAs, applicationId & ledgerId are included in the log context

3ad9cd58d9869ce9caaaeca17608b09091cba695

- [JSON API] The template id & choice name (if present) are now logged on command submissions in the Json API (at trace level)

3583fd1885e3676867361a220b77c82d4ed1962c

- [JSON-API] The source and the path for incoming http requests are now logged
- [JSON-API] The http response for a request is now logged

3d2d4b969695339017f6fb11a44a7dd36ef2b2a5

- [JSON API] Errors which arise from the CommandService are not logged twice anymore (thus reducing noise)

603a4a4b424493cdc71b9ed8d231c09fe016eb88

- [JSON API] log statements now include the date next to the time

de7a462082fc69e2c6996b2e5ac4c05f5c790f09

- [HTTP-JSON] Calls which trigger a submitAndWaitRequest are logged with the command id provided in the log ctx

13767b5c5d531690a0d4bb7e77cf2a66ab9ac3d3

[HTTP-JSON]
- fixed that json log output could not be enabled via cli options except via usage of env vars

4037b1cd9e4fce053370acfee72430ac92500a04

[HTTP/JSON API]
- metrics reporting can now be chosen via the command line option --metrics-reporter (currently hidden), valid options are console, csv://, graphite:// and prometheus://
- metrics reporting interval can also now be chosen via a command line option --metrics-reporting-interval (currently hidden)

beb2d138f2939f3d9cfbb29cdf6a607f9df6da2b

- [Ledger HTTP Json Service] Logging output can now be in JSON either via providing the cli option `--log-encoder json` or via setting the env var `LOG_FORMAT_JSON=true`

4f0ceff7a2ad2865087250a7f98e17e8fd7e8f75

http-json:
- add contextual id as logging context to distinguish different application runs in logs
- add request id as logging context to distinguish different http requests within an application run
- add for non-static endpoints trace logs which show how long processing it took in ns

112c387e5a4c82a2ce78072827c4ace44dd8305e

- [Ledger HTTP Json Service] Logging now also tells service name if log level was changed.

```

**Comment by stefanobaghino-da** [comment]
@realvictorprm What is still needed for this to be closed?

**Comment by realvictorprm** [comment]
With #11768 being already merged we can probably close this already. The only PR which is not merged yet and somewhat related to this is #11809 which forward-ports metrics which have been introduced in the process of solving this issue.

**Comment by stefanobaghino-da** [comment]
@realvictorprm Since #11809 has been closed I'll close this. Re-open if needed.
